<a href="https://colab.research.google.com/github/gbodra/mestrado-ia-pel208-deep-learning/blob/main/DeepLearning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import time
import os
from datetime import datetime

(train_x, train_y), (test_x, test_y) = keras.datasets.mnist.load_data()
train_x = train_x / 255.0
test_x = test_x / 255.0

train_x = tf.expand_dims(train_x, 3)
test_x = tf.expand_dims(test_x, 3)

val_x = train_x[:5000]
val_y = train_y[:5000]

In [2]:
lenet_5_model = keras.models.Sequential([
    keras.layers.Conv2D(6, kernel_size=5, strides=1,  activation='tanh', input_shape=train_x[0].shape, padding='same'),
    keras.layers.AveragePooling2D(),
    keras.layers.Conv2D(16, kernel_size=5, strides=1, activation='tanh', padding='valid'),
    keras.layers.AveragePooling2D(),
    keras.layers.Flatten(),
    keras.layers.Dense(120, activation='tanh'),
    keras.layers.Dense(84, activation='tanh'),
    keras.layers.Dense(10, activation='softmax')
])

In [3]:
lenet_5_model.compile(optimizer='adam', loss=keras.losses.sparse_categorical_crossentropy, metrics=['accuracy'])

In [4]:
start_time = datetime.timestamp(datetime.now())
lenet_5_model.fit(train_x, train_y, epochs=100, validation_data=(val_x, val_y))
end_time = datetime.timestamp(datetime.now())
elapsed = end_time - start_time
elapsed

Epoch 1/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2236 - accuracy: 0.9337 - val_loss: 0.0874 - val_accuracy: 0.9768
Epoch 2/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0850 - accuracy: 0.9744 - val_loss: 0.0522 - val_accuracy: 0.9836
Epoch 3/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0581 - accuracy: 0.9823 - val_loss: 0.0355 - val_accuracy: 0.9910
Epoch 4/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0433 - accuracy: 0.9867 - val_loss: 0.0309 - val_accuracy: 0.9910
Epoch 5/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0363 - accuracy: 0.9882 - val_loss: 0.0273 - val_accuracy: 0.9916
Epoch 6/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0286 - accuracy: 0.9908 - val_loss: 0.0262 - val_accuracy: 0.9914
Epoch 7/100
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0238 - accuracy: 0.9925 - val_loss: 0.0185 - val_ac

426.4237208366394

In [5]:
lenet_5_model.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0657 - accuracy: 0.9885


[0.06565392762422562, 0.9884999990463257]

In [6]:
!pip install git+git://github.com/albertbup/deep-belief-network.git@master_gpu

  Cloning git://github.com/albertbup/deep-belief-network.git (to revision master_gpu) to /tmp/pip-req-build-p8u6rrf4
  Running command git clone -q git://github.com/albertbup/deep-belief-network.git /tmp/pip-req-build-p8u6rrf4
  Running command git checkout -b master_gpu --track origin/master_gpu
  Switched to a new branch 'master_gpu'
  Branch 'master_gpu' set up to track remote branch 'master_gpu' from 'origin'.
     |████████████████████████████████| 320.4MB 57kB/s 
  Created wheel for deep-belief-network: filename=deep_belief_network-1.0.3-cp36-none-any.whl size=13464 sha256=3bb7288b397d80686419bb4c18c8312bc02d63f176c090b94c52c56bb0d7112d
  Stored in directory: /tmp/pip-ephem-wheel-cache-2ovwl8ae/wheels/fe/5a/c7/18fc337ca5590a8b7c74c17ddba7a63935ce7a5bafb6cdec02
Successfully built deep-belief-network


In [2]:
import numpy as np

np.random.seed(0)
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics.classification import accuracy_score
from datetime import datetime

start_time = datetime.timestamp(datetime.now())

# from dbn.tensorflow import SupervisedDBNClassification
from dbn import SupervisedDBNClassification
# use "from dbn import SupervisedDBNClassification" for computations on CPU with numpy

# Loading dataset
digits = load_digits()
X, Y = digits.data, digits.target

# Data scaling
X = (X / 16).astype(np.float32)

# Splitting data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

# Training
classifier = SupervisedDBNClassification(hidden_layers_structure=[256, 256],
                                         learning_rate_rbm=0.05,
                                         learning_rate=0.1,
                                         n_epochs_rbm=10,
                                         n_iter_backprop=100,
                                         batch_size=32,
                                         activation_function='relu',
                                         dropout_p=0.2)
classifier.fit(X_train, Y_train)

# Save the model
classifier.save('model.pkl')

# Restore it
classifier = SupervisedDBNClassification.load('model.pkl')

# Test
Y_pred = classifier.predict(X_test)
print('Done.\nAccuracy: %f' % accuracy_score(Y_test, Y_pred))

end_time = datetime.timestamp(datetime.now())
elapsed = end_time - start_time
elapsed

[START] Pre-training step:
>> Epoch 1 finished 	RBM Reconstruction error 2.710084
>> Epoch 2 finished 	RBM Reconstruction error 1.796745
>> Epoch 3 finished 	RBM Reconstruction error 1.434709
>> Epoch 4 finished 	RBM Reconstruction error 1.201816
>> Epoch 5 finished 	RBM Reconstruction error 1.108185
>> Epoch 6 finished 	RBM Reconstruction error 0.992179
>> Epoch 7 finished 	RBM Reconstruction error 0.916095
>> Epoch 8 finished 	RBM Reconstruction error 0.919944
>> Epoch 9 finished 	RBM Reconstruction error 0.853218
>> Epoch 10 finished 	RBM Reconstruction error 0.850179
>> Epoch 1 finished 	RBM Reconstruction error 2.870218
>> Epoch 2 finished 	RBM Reconstruction error 1.499493
>> Epoch 3 finished 	RBM Reconstruction error 1.111499
>> Epoch 4 finished 	RBM Reconstruction error 0.950356
>> Epoch 5 finished 	RBM Reconstruction error 0.802263
>> Epoch 6 finished 	RBM Reconstruction error 0.677447
>> Epoch 7 finished 	RBM Reconstruction error 0.610802
>> Epoch 8 finished 	RBM Reconstructi

103.4248399734497

# Análise dos resultados
## LeNet-5


|Epochs	| Tempo de treinamento | Perda | Acuracidade | GPU |
|-------|----------------------|-------|-------------|-----|
| 5     | 206s                 | 0.0567| 0.9810      | Não |
| 5     | 21s                  | 0.0392| 0.9876      | Sim |
| 10    | 416s                 | 0.0497| 0.9855      | Não |
| 50    | 212s                 | 0.0626| 0.9859      | Sim |
| 100   | 426s                 | 0.0657| 0.9885      | Sim |

## DBN

|Epochs RBM | Epochs ANN | Tempo de Treinamento | Perda | Acuracidade |
|-----------|------------|----------------------|-------|-------------|
| 20        | 100        | 103s                 | 2.2227| 0.9805      |

## Conclusão

Os tempos de execução são muito próximos quando são utilizadas GPUs para realizar o processamento, porém a Lenet-5 produz resultados mais consistentes em menos iterações.